### PR TITLE
HLD for Dump Utility Added

### PR DESCRIPTION
**Pull request by vivekrnv**
Signed-off-by: Vivek Reddy Karri <vkarri@nvidia.com>

The high-level design document for the `dump` cli utility

### CODE REVIEW

File: doc/Dump-Utility.md [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+* Case 1: field and value in the MatchRequest are None. Result: MatchEngine returns all the keys which are regex matched by "Table<table_sep>key_regex".
+* Case 2: field and value in the MatchRequest are not None and a set of keys are matched by the "Table|key_regex". Result: The MatchEngine looks into each of these keys and returns those keys who has their field-value pairs equated to what is provided. 
+* Case 3: For a valid combination of db/file, Table, key_regex, field and value, if all the field-value pairs are required, set just_keys to true.
+* Case 4: For a valid combination of db/file, Table, key_regex, field and value, if only a few specific fields are required, set just_keys to false and use return_fields option.
+
+###### JSON Template 3: Return Dictionary by the MatchEngine:
+
+```
+{  
+  "error": "<STR>",             # Error String, if any. Empty Otherwise
```
Comment (nazariig) [inline doc/Dump-Utility.md:688]: @vivekreddynv can we have some examples with this attribute? Also it lacks an explanation of possible use cases.

File: doc/Dump-Utility.md [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+```
+root@sonic# dump state --help
+Usage: dump state [OPTIONS] MODULE ARG
+
+  Dump the redis state of the module specified
+
+Options:
+  -s, --show     Display all the Modules and the name of their correponding secondary argument 
+  -d, --db TEXT  Only dump from these Databases
+  -t, --table    Print in tabular format  [default: False]
```
Comment (nazariig) [inline doc/Dump-Utility.md]: @vivekreddynv please provide some CLI examples for general usage and for different options, e.g.:
```
root@sonic# dump state copp Ethernet0
...

root@sonic# dump state copp Ethernet0 --table
...

root@sonic# dump state copp Ethernet0 --rid
...

root@sonic# dump state copp Ethernet0 --key-map
...

root@sonic# dump state copp Ethernet0 --verbose
...
```

File: doc/Dump-Utility.md [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+  -r, --rid      Dont Extract VidToRid Mappings for ASIC DB Dumps  [default: False]
+  -k, --key-map  Only Return Key mappings, Dont return Field-Value Dumps [default: False]
+  -v, --verbose  Prints any intermediate output to stdout useful for dev & troubleshooting  [default: False]
+  --help         Show this message and exit.
+  
+root@sonic# dump state --show
+Module            Args
```
Comment (nazariig) [inline doc/Dump-Utility.md]: @vivekreddynv `Args` is not a best identifier here. Can we have something more meaningful instead? e.g., `items`, `entities`, `objects`, etc.?

File: doc/Dump-Utility.md [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+--------------    ------
+copp              trap_id
```
Comment (nazariig) [inline doc/Dump-Utility.md]: @vivekreddynv what will be the output if `Module` supports several `Arguments`? Can we add some examples?

File: doc/Dump-Utility.md [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+This utility provides the base infrastructure and guidelines to make is easy for the developers to extend and support the utility for different modules.
+
+## **1 Requirements**
+
+### 1.1 Intended Usage
+
+Note: Currently, `state` is the only subcommand under `dump`.
+
+```
+dump state <feature/module> <arg>
```
Comment (nazariig) [inline doc/Dump-Utility.md]: @vivekreddynv the description seems to be invalid: `dump state <feature/module> <args | all>`

File: doc/Dump-Utility.md [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+   ├──── etc/
+   ├──── log/  
+   ├──── ......
+   ├──── ......
+```
+
+
+
+
+
```
Comment (nazariig) [inline doc/Dump-Utility.md]: @vivekreddynv please remove extra lines

File: doc/Dump-Utility.md (hunk 4 above)
Comment (vivekkarri) [reply doc/Dump-Utility.md]: Currently we only support a single argument case.

File: doc/Dump-Utility.md (hunk 1 above)
Comment (vivekkarri) [reply doc/Dump-Utility.md:688]: Thanks for the suggestion. Added the possible error strings thrown

File: doc/Dump-Utility.md (hunk 2 above)
Comment (vivekkarri) [reply doc/Dump-Utility.md]: Done!, Thanks for the input

File: doc/Dump-Utility.md (hunk 3 above)
Comment (vivekkarri) [reply doc/Dump-Utility.md]: Done

File: doc/Dump-Utility.md (hunk 5 above)
Comment (vivekkarri) [reply doc/Dump-Utility.md]: Done!

File: doc/Dump-Utility.md (hunk 2 above)
Comment (vivekkarri) [reply doc/Dump-Utility.md]: Only, the tabular format is left.

Comment (vivekkarri) [reply doc/Dump-Utility.md]: Tabular format output completed

**Review by nazariig — APPROVED** [review]

**Pull request closed without merging**